assets/src/demo/app.js

```diff
@@ -0,0 +1,188 @@
+import { TextControl, Button, Spinner } from '@wordpress/components';
+import { Fragment, Component, render as domRender } from '@wordpress/element';
+import { withState, compose } from '@wordpress/compose';
+import { withSelect, withDispatch } from '@wordpress/data';
+import { isModelEntityOfModel } from '@eventespresso/validators';
+import './app.css';
+
+class EventInput extends Component {
+	constructor( props ) {
+		super( props );
+		this.state = { eventId: props.eventId || 0 };
+	}
+
+	setEventId = ( id ) => {
+		this.setState( { eventId: id } );
+	};
+
+	onEventSubmit = () => this.props.onEventClick( this.state.eventId );
+
+	render() {
+		return <div>
+			<TextControl
+				label={ 'Enter the EventId here' }
+				value={ this.state.eventId }
+				onChange={ this.setEventId }
+			/>
+			<Button
+				label={ 'Retrieve Event' }
+				isPrimary={ true }
+				onClick={ this.onEventSubmit }
+			>
+				Retrieve Event
+			</Button>
+		</div>;
+	}
+}
+
+const DatesList = ( { datetimes, onDateClone } ) => {
+	function getDatetimesList() {
+		if ( datetimes.length > 0 ) {
+			return datetimes.map( ( datetime ) => <li key={ datetime.id }>
+				{ datetime.start.toFormat() }
+				<br />
+				<CloneDateTimeButton
+					datetime={ datetime }
+					onCloneDateTime={ onDateClone }
+				/>
+			</li> );
+		}
+		return <li><Spinner /></li>;
+	}
+	return <div>
+		<h2>Datetimes List</h2>
+		<ul>{ getDatetimesList() }</ul>
+	</div>;
+};
+
+const DatetimesList = withDispatch( ( dispatch, ownProps, { select } ) => {
+	async function onDateClone( datetime ) {
+		const { getRelatedEntities } = select( 'eventespresso/core' );
+		const { createEntity, createRelations } = dispatch( 'eventespresso/core' );
+		const tickets = getRelatedEntities( datetime, 'tickets' );
+		const newDateTime = await createEntity( 'datetime', datetime.forClone );
+		createRelations(
+			'datetime',
+			newDateTime.id,
+			'tickets',
+			tickets
+		);
+		createRelations(
+			'event',
+			ownProps.eventId,
+			'datetimes',
+			[ newDateTime ]
+		);
+	}
+	return { onDateClone };
+} )( DatesList );
```

Comment:
Note how DatesList (which is responsible for outputting the list of datetimes and the clone button) has the `withDispatch` HOC wrapping it.  This HOC primarily takes care of the clone button behaviour which when triggered:

* Retrieves related tickets on the datetime clicked.
* clones the new Datetime entity from the one clicked and creates the relation to tickets on that new entity
* creates the relation to the event id on that new entity (I _could_ have also just retrieved the event from the store state if I wanted to).

Note, this does not feed the `datetimes` array to the component as that will automatically get updated further upstream.